### PR TITLE
add pagination to list_measurments()

### PR DESCRIPTION
**Pull request by tuix**
# Description

Include a summary of this pull request

New dependencies: (Added to requirements_*.txt)

Fixes # (potential issues that are fixed by this PR)

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue -> bump patch)
- [ ] New feature (non-breaking change which adds functionality -> bump minor version)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected -> bump major version)
- [ ] Documentation Update
- [ ] CI/CD workflows Update

## Checklist

- [ ] I have added tests that prove that my fix/feature works
- [ ] Linters pass locally and I have followed PEP8 code style
- [ ] New and existing tests pass locally
- [ ] I have updated the documentation if needed
- [ ] I have commented hard-to-understand areas in the code

## Requirements

- [ ] I have updated the MVG version

### CODE REVIEW

File: mvg/mvg.py [hunk 1]
```diff
@@ -573,7 +588,11 @@ def read_single_measurement(self, sid: str, timestamp: int) -> dict:
         return response.json()
 
     def list_tabular_measurements(
-        self, sid: str, start_timestamp: int = None, end_timestamp: int = None
+        self,
+        sid: str,
+        start_timestamp: int = None,
+        end_timestamp: int = None,
+        list_all=True,
```
Comment (claezon) [inline mvg/mvg.py:595]: We should also incorporate the pagination in this method to enable it for the clients (/Pipeline). Also, I suggest that we always default to a limit.

File: mvg/mvg.py [hunk 2]
```diff
@@ -534,10 +538,21 @@ def list_measurements(self, sid: str) -> list:
...
 
-        response = self._request("get", f"/sources/{sid}/measurements")
-        all_measurements = response.json()
-
-        logger.info("%s measurements in database", len(all_measurements))
+        url = f"/sources/{sid}/measurements"
+        response = self._request("get", url)
+        resp_first = response.json()
+        all_measurements = resp_first["items"]
+        num_meaurements = resp_first["total"]
```
Comment (claezon) [inline mvg/mvg.py:545]: Typo (num_measurements).

**Review by claezon — CHANGES_REQUESTED** [review]
A proposal for an overall change regarding this PR to enable pagination.

File: mvg/mvg.py [hunk 3]
```diff
@@ -518,13 +518,17 @@ def create_tabular_measurement(
         )
 
     # in example
-    def list_measurements(self, sid: str) -> list:
-        """Retrieves all measurements (all timestamps and metadata) for a source.
+    def list_measurements(self, sid: str, list_all=True) -> list:
```
[the remark this answers is not in the document]
Comment (tuix) [reply mvg/mvg.py:521]: I am not sure whether pipeline would require the pagination at this time. This was done this way to keep it backwards compatible where this method would return all results. If we want to incorporate pagination here that would require update in jupyter notebooks that use this and I wanted this to be merged ASAP. If you don't mind we can do it this way for now and I can create a separate issue for introducing pagination later.

[the remark this answers is not in the document]
Comment (claezon) [reply mvg/mvg.py:521]: Yes, I suspected that was the reason. But we could still let the client use the pagination parameters, but the response would still be an array (leaving out the pagination meta: offset, limit, total...)? Later, in connection with implementing versioning of the API(s), we could replace the response array with a uniformed response object.

[the remark this answers is not in the document]
Comment (tuix) [reply mvg/mvg.py]: I added offset and limit to both "list_*" methods. by default it returns all measuremnts

File: mvg/mvg.py [hunk 4]
```diff
@@ -533,11 +539,32 @@ def list_measurements(self, sid: str) -> list:
...
+            for idx in range(1, num_reqs + 1):
+                offset = idx * limit
+                response = self._request("get", url, params={"offset": offset})
+                all_measurements += response.json()["items"]
+        else:
+            params = {}
+            if limit is not None:
+                params["limit"] = limit
+            if offset is not None:
+                params["offset"] = offset
```
Comment (claezon) [inline mvg/mvg.py:560]: A suggestion to reduce the level of nested if statements is to move the declaration of params to the beginning:

```python
    params = {}
    if limit is not None:
        params["limit"] = limit
    if offset is not None:
        params["offset"] = offset

    if not params:
        # List all by default
    else:
        # Use pagination
```

File: mvg/mvg.py [hunk 5]
```diff
@@ -588,30 +620,54 @@ def list_tabular_measurements(
...
-        return response.json()
+        resp_first = response.json()
+        all_measurements = resp_first["items"]
+        num_meaurements = resp_first["total"]
+        if (
+            start_timestamp is None
+            and end_timestamp is None
+            and offset is None
+            and limit is None
+        ):
```
Comment (claezon) [inline mvg/mvg.py:662]: I believe checking the `query_params` would be enough.

File: mvg/mvg.py [hunk 6]
```diff
@@ -588,30 +620,54 @@ def list_tabular_measurements(
...
-        KPI, and metadata corresponding to a measurement
+        column names, and metadata corresponding to a measurement.
         """
         logger.info("endpoint %s", self.endpoint)
         logger.info("retrieving all measurements from source id=%s", sid)
 
-        query_params_list = []
-        query_params_str = ""
+        url = f"/sources/{sid}/measurements/tabular"
+        query_params = {}
```
Comment (claezon) [inline mvg/mvg.py:637]: The name `params` is used in `list_measurements()`. Use the same name?

File: mvg/mvg.py [hunk 7]
```diff
@@ -588,30 +620,54 @@ def list_tabular_measurements(
...
-            "get", f"/sources/{sid}/measurements/tabular{query_params_str}"
+            "get",
+            url,
+            params=query_params,
         )
 
-        return response.json()
+        resp_first = response.json()
+        all_measurements = resp_first["items"]
+        num_meaurements = resp_first["total"]
```
Comment (claezon) [inline mvg/mvg.py:656]: Typo (num_measurements).

**Review by claezon — CHANGES_REQUESTED** [review]
Nice Mohsen! A few small suggestions that you can choose to ignore and a typo.

**Review by claezon — APPROVED** [review]
Nice work Mohsen!

**Pull request merged**